File: dashboard/scripts/validate_production.js
#!/usr/bin/env node

/**
 * Production Build Validation Script for Angular Frontend
 * 
 * This script validates that the Angular frontend is properly configured for production
 * and contains no mock, seed, or sample data.
 */

const fs = require('fs');
const path = require('path');
const glob = require('glob');

class FrontendProductionValidator {
    constructor() {
        this.errors = [];
        this.warnings = [];
        this.projectRoot = path.resolve(__dirname, '..');
    }

    /**
     * Run all validation checks
     */
    validateAll() {
        console.log('🔍 Starting production validation for Angular frontend...');
        
        // Configuration validation
        this.validateAngularConfig();
        this.validateEnvironmentConfig();
        
        // Code validation
        this.validateNoMockData();
        this.validateNoSeedData();
        this.validateNoDebugCode();
        
        // Build validation
        this.validateBuildOutput();
        
        // Security validation
        this.validateSecuritySettings();
        
        return {
            status: this.errors.length === 0 ? 'PASS' : 'FAIL',
            errors: this.errors,
            warnings: this.warnings,
            checksPerformed: [
                'Angular configuration check',
                'Environment configuration check',
                'Mock data detection',
                'Seed data detection',
                'Debug code detection',
                'Build output validation',
                'Security settings validation'
            ]
        };
    }

    /**
     * Validate Angular configuration
     */
    validateAngularConfig() {
        console.log('  ✓ Validating Angular configuration...');
        
        const angularJsonPath = path.join(this.projectRoot, 'angular.json');
        if (!fs.existsSync(angularJsonPath)) {
            this.errors.push('angular.json file not found');
            return;
        }

        try {
            const angularConfig = JSON.parse(fs.readFileSync(angularJsonPath, 'utf8'));
            const buildConfig = angularConfig.projects?.dashboard?.architect?.build;
            
            if (!buildConfig) {
                this.errors.push('Build configuration not found in angular.json');
                return;
            }

            // Check production configuration
            const prodConfig = buildConfig.configurations?.production;
            if (!prodConfig) {
                this.errors.push('Production configuration not found in angular.json');
                return;
            }

            // Validate production settings
            if (prodConfig.optimization !== true) {
                this.warnings.push('Optimization is not enabled for production build');
            }

            if (prodConfig.sourceMap !== false) {
                this.warnings.push('Source maps are enabled for production build');
            }

            if (prodConfig.extractLicenses !== true) {
                this.warnings.push('License extraction is not enabled for production build');
            }

        } catch (error) {
            this.errors.push(`Failed to parse angular.json: ${error.message}`);
        }
    }

    /**
     * Validate environment configuration
     */
    validateEnvironmentConfig() {
        console.log('  ✓ Validating environment configuration...');
        
        const envProdPath = path.join(this.projectRoot, 'src/environments/environment.prod.ts');
        const envPath = path.join(this.projectRoot, 'src/environments/environment.ts');
        
        if (!fs.existsSync(envProdPath)) {
            this.warnings.push('Production environment file not found');
        } else {
            try {
                const envContent = fs.readFileSync(envProdPath, 'utf8');
                
                if (envContent.includes('production: false')) {
                    this.errors.push('Production flag is set to false in production environment');
                }
                
                if (envContent.includes('localhost') || envContent.includes('127.0.0.1')) {
                    this.warnings.push('Production environment contains localhost URLs');
                }
                
            } catch (error) {
                this.warnings.push(`Could not read production environment file: ${error.message}`);
            }
        }
        
        // Check development environment for production issues
        if (fs.existsSync(envPath)) {
            try {
                const envContent = fs.readFileSync(envPath, 'utf8');
                
                if (envContent.includes('production: true')) {
                    this.warnings.push('Development environment has production flag set to true');
                }
                
            } catch (error) {
                this.warnings.push(`Could not read development environment file: ${error.message}`);
            }
        }
    }

    /**
     * Check for mock data in the codebase
     */
    validateNoMockData() {
        console.log('  ✓ Checking for mock data...');
        
        const mockPatterns = [
            /mock[_-]?data/gi,
            /MockData/g,
            /MOCK_.*=/g,
            /fake[_-]?data/gi,
            /FakeData/g,
            /test[_-]?data.*=/gi,
            /sample[_-]?data/gi,
            /SampleData/g,
            /dummy[_-]?data/gi,
            /DummyData/g
        ];
        
        this._scanFilesForPatterns(['**/*.ts', '**/*.js'], mockPatterns, 'mock data');
    }

    /**
     * Check for seed data in the codebase
     */
    validateNoSeedData() {
        console.log('  ✓ Checking for seed data...');
        
        const seedPatterns = [
            /seed[_-]?data/gi,
            /SeedData/g,
            /initial[_-]?data/gi,
            /InitialData/g,
            /bootstrap[_-]?data/gi,
            /BootstrapData/g,
            /fixtures/gi
        ];
        
        this._scanFilesForPatterns(['**/*.ts', '**/*.js'], seedPatterns, 'seed data');
        
        // Check for JSON files that might contain seed data
        const dataFiles = glob.sync('**/*.json', { 
            cwd: this.projectRoot,
            ignore: ['node_modules/**', 'dist/**', 'package*.json', 'angular.json', 'tsconfig*.json']
        });
        
        dataFiles.forEach(file => {
            this.warnings.push(`Found data file that might contain seed data: ${file}`);
        });
    }

    /**
     * Check for debug code that shouldn't be in production
     */
    validateNoDebugCode() {
        console.log('  ✓ Checking for debug code...');
        
        const debugPatterns = [
            /console\.log\(/g,
            /console\.debug\(/g,
            /console\.warn\(/g,
            /console\.error\(/g,
            /debugger;/g,
            /TODO:/gi,
            /FIXME:/gi,
            /XXX:/gi,
            /alert\(/g,
            /confirm\(/g
        ];
        
        this._scanFilesForPatterns(['**/*.ts', '**/*.js'], debugPatterns, 'debug code', true);
    }

    /**
     * Validate build output
     */
    validateBuildOutput() {
        console.log('  ✓ Validating build output...');
        
        const distPath = path.join(this.projectRoot, 'dist');
        if (!fs.existsSync(distPath)) {
            this.warnings.push('Build output directory (dist) not found - run ng build first');
            return;
        }

        const dashboardPath = path.join(distPath, 'dashboard');
        if (!fs.existsSync(dashboardPath)) {
            this.warnings.push('Dashboard build output not found in dist directory');
            return;
        }

        // Check for index.html
        const indexPath = path.join(dashboardPath, 'index.html');
        if (!fs.existsSync(indexPath)) {
            this.errors.push('index.html not found in build output');
        }

        // Check for main bundle files
        const files = fs.readdirSync(dashboardPath);
        const hasMainJs = files.some(file => file.startsWith('main.') && file.endsWith('.js'));
        const hasPolyfillsJs = files.some(file => file.startsWith('polyfills.') && file.endsWith('.js'));
        
        if (!hasMainJs) {
            this.errors.push('Main JavaScript bundle not found in build output');
        }
        
        if (!hasPolyfillsJs) {
            this.warnings.push('Polyfills JavaScript bundle not found in build output');
        }

        // Check for source maps in production build
        const sourceMapFiles = files.filter(file => file.endsWith('.map'));
        if (sourceMapFiles.length > 0) {
            this.warnings.push(`Source map files found in production build: ${sourceMapFiles.join(', ')}`);
        }
    }

    /**
     * Validate security settings
     */
    validateSecuritySettings() {
        console.log('  ✓ Validating security settings...');
        
        // Check for security headers in nginx config
        const nginxConfigPath = path.join(this.projectRoot, 'nginx.conf');
        if (fs.existsSync(nginxConfigPath)) {
            try {
                const nginxContent = fs.readFileSync(nginxConfigPath, 'utf8');
                
                const securityHeaders = [
                    'X-Frame-Options',
                    'X-XSS-Protection',
                    'X-Content-Type-Options',
                    'Content-Security-Policy'
                ];
                
                securityHeaders.forEach(header => {
                    if (!nginxContent.includes(header)) {
                        this.warnings.push(`Security header ${header} not found in nginx configuration`);
                    }
                });
                
            } catch (error) {
                this.warnings.push(`Could not read nginx configuration: ${error.message}`);
            }
        } else {
            this.warnings.push('nginx.conf not found - security headers may not be configured');
        }
    }

    /**
     * Scan files for specific patterns
     */
    _scanFilesForPatterns(filePatterns, patterns, description, isWarning = false) {
        const excludePatterns = [
            'node_modules/**',
            'dist/**',
            '.angular/**',
            'coverage/**',
            '**/*.spec.ts',
            '**/*.test.ts',
            '**/test-setup.ts',
            '**/validate_production.js'
        ];

        filePatterns.forEach(filePattern => {
            const files = glob.sync(filePattern, {
                cwd: this.projectRoot,
                ignore: excludePatterns
            });

            files.forEach(file => {
                try {
                    const content = fs.readFileSync(path.join(this.projectRoot, file), 'utf8');
                    
                    patterns.forEach(pattern => {
                        const matches = content.match(pattern);
                        if (matches) {
                            matches.forEach(match => {
                                const lines = content.substring(0, content.indexOf(match)).split('\n');
                                const lineNum = lines.length;
                                const message = `Found ${description} in ${file}:${lineNum} - '${match}'`;
                                
                                if (isWarning) {
                                    this.warnings.push(message);
                                } else {
                                    this.errors.push(message);
                                }
                            });
                        }
                    });
                    
                } catch (error) {
                    this.warnings.push(`Could not scan ${file}: ${error.message}`);
                }
            });
        });
    }
}

/**
 * Main validation function
 */
function main() {
    const validator = new FrontendProductionValidator();
    const results = validator.validateAll();
    
    console.log('\n' + '='.repeat(60));
    console.log('🏁 FRONTEND PRODUCTION VALIDATION RESULTS');
    console.log('='.repeat(60));
    
    console.log(`\n📊 Status: ${results.status}`);
    console.log(`🔍 Checks performed: ${results.checksPerformed.length}`);
    
    if (results.errors.length > 0) {
        console.log(`\n❌ Errors (${results.errors.length}):`);
        results.errors.forEach(error => {
            console.log(`  • ${error}`);
        });
    }
    
    if (results.warnings.length > 0) {
        console.log(`\n⚠️  Warnings (${results.warnings.length}):`);
        results.warnings.forEach(warning => {
            console.log(`  • ${warning}`);
        });
    }
    
    if (results.errors.length === 0 && results.warnings.length === 0) {
        console.log('\n✅ All validation checks passed!');
    }
    
    // Save results to file
    const resultsFile = path.join(__dirname, 'validation_results.json');
    fs.writeFileSync(resultsFile, JSON.stringify(results, null, 2));
    
    console.log(`\n📄 Detailed results saved to: ${resultsFile}`);
    
    // Exit with appropriate code
    process.exit(results.status === 'PASS' ? 0 : 1);
}

if (require.main === module) {
    main();
}

File: middleware/scripts/validate_production.py
#!/usr/bin/env python3
"""
Production Build Validation Script for FastAPI Middleware

This script validates that the FastAPI middleware is properly configured for production
and contains no mock, seed, or sample data.
"""

import os
import sys
import re
import json
import asyncio
from pathlib import Path
from typing import List, Dict, Any, Tuple

# Add the middleware project to the path
sys.path.insert(0, os.path.dirname(__file__).parent)


class MiddlewareProductionValidator:
    """Validates FastAPI middleware for production readiness."""
    
    def __init__(self):
        self.errors = []
        self.warnings = []
        self.project_root = Path(__file__).parent.parent
        
    def validate_all(self) -> Dict[str, Any]:
        """Run all validation checks."""
        print("🔍 Starting production validation for FastAPI middleware...")
        
        # Configuration validation
        self.validate_environment_config()
        
        # Code validation
        self.validate_no_mock_data()
        self.validate_no_seed_data()
        self.validate_no_debug_code()
        
        # Security validation
        self.validate_security_settings()
        
        # Performance validation
        self.validate_performance_config()
        
        return {
            "status": "PASS" if not self.errors else "FAIL",
            "errors": self.errors,
            "warnings": self.warnings,
            "checks_performed": [
                "Environment configuration check",
                "Mock data detection",
                "Seed data detection", 
                "Debug code detection",
                "Security settings validation",
                "Performance configuration check"
            ]
        }
    
    def validate_environment_config(self):
        """Validate environment configuration."""
        print("  ✓ Validating environment configuration...")
        
        # Check environment variables
        required_vars = [
            'DATABASE_URL',
            'REDIS_URL',
            'DJANGO_BACKEND_URL'
        ]
        
        for var in required_vars:
            if not os.getenv(var):
                self.warnings.append(f"Environment variable {var} is not set")
        
        # Check FASTAPI_ENV
        fastapi_env = os.getenv('FASTAPI_ENV', 'development')
        if fastapi_env != 'production':
            self.warnings.append(f"FASTAPI_ENV is set to '{fastapi_env}', should be 'production'")
        
        # Check LOG_LEVEL
        log_level = os.getenv('LOG_LEVEL', 'debug')
        if log_level.lower() in ['debug', 'trace']:
            self.warnings.append(f"LOG_LEVEL is set to '{log_level}', consider 'info' or 'warning' for production")
    
    def validate_no_mock_data(self):
        """Check for mock data in the codebase."""
        print("  ✓ Checking for mock data...")
        
        mock_patterns = [
            r'mock_data',
            r'MockData',
            r'MOCK_.*=',
            r'fake_data',
            r'FakeData',
            r'test_data.*=',
            r'sample_data',
            r'SampleData',
            r'dummy_data',
            r'DummyData',
            r'Mock\w+\(',
            r'@mock\.',
            r'unittest\.mock'
        ]
        
        self._scan_files_for_patterns(mock_patterns, "mock data")
    
    def validate_no_seed_data(self):
        """Check for seed data in the codebase."""
        print("  ✓ Checking for seed data...")
        
        seed_patterns = [
            r'seed_data',
            r'SeedData',
            r'initial_data',
            r'InitialData',
            r'bootstrap_data',
            r'BootstrapData',
            r'sample_.*_data',
            r'example_.*_data'
        ]
        
        self._scan_files_for_patterns(seed_patterns, "seed data")
        
        # Check for data files that might contain seed data
        data_files = list(self.project_root.glob('**/*.json')) + \
                    list(self.project_root.glob('**/*.yaml')) + \
                    list(self.project_root.glob('**/*.yml'))
        
        for data_file in data_files:
            # Skip configuration files and package files
            if data_file.name in ['package.json', 'pyproject.toml', 'requirements.txt']:
                continue
            
            # Skip files in excluded directories
            excluded_dirs = {'__pycache__', '.git', 'venv', 'env', 'node_modules', '.pytest_cache'}
            if any(excluded_dir in data_file.parts for excluded_dir in excluded_dirs):
                continue
            
            self.warnings.append(f"Found data file that might contain seed data: {data_file}")
    
    def validate_no_debug_code(self):
        """Check for debug code that shouldn't be in production."""
        print("  ✓ Checking for debug code...")
        
        debug_patterns = [
            r'print\s*\(',
            r'pdb\.set_trace\(\)',
            r'breakpoint\(\)',
            r'import\s+pdb',
            r'console\.log\(',
            r'debugger;',
            r'TODO:',
            r'FIXME:',
            r'XXX:',
            r'logger\.debug\(',
            r'logging\.DEBUG'
        ]
        
        self._scan_files_for_patterns(debug_patterns, "debug code", is_warning=True)
    
    def validate_security_settings(self):
        """Validate security-related settings."""
        print("  ✓ Validating security settings...")
        
        # Check main.py for security configurations
        main_file = self.project_root / 'main.py'
        if main_file.exists():
            try:
                with open(main_file, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                # Check for CORS configuration
                if 'CORSMiddleware' in content:
                    if 'allow_origins=["*"]' in content:
                        self.errors.append("CORS is configured to allow all origins (*) - security risk")
                    if 'allow_credentials=True' in content and 'allow_origins=["*"]' in content:
                        self.errors.append("CORS allows credentials with wildcard origins - security risk")
                
                # Check for docs in production
                if 'docs_url=' not in content or 'docs_url=None' not in content:
                    self.warnings.append("API documentation might be exposed in production")
                
            except Exception as e:
                self.warnings.append(f"Could not analyze main.py: {str(e)}")
    
    def validate_performance_config(self):
        """Validate performance-related configuration."""
        print("  ✓ Validating performance configuration...")
        
        # Check for performance monitoring
        performance_files = list(self.project_root.glob('**/performance/*.py'))
        if not performance_files:
            self.warnings.append("No performance monitoring modules found")
        
        # Check for caching configuration
        cache_files = list(self.project_root.glob('**/caching/*.py'))
        if not cache_files:
            self.warnings.append("No caching modules found")
        
        # Check Redis configuration
        redis_url = os.getenv('REDIS_URL')
        if not redis_url:
            self.warnings.append("Redis URL not configured - caching may not work")
    
    def _scan_files_for_patterns(self, patterns: List[str], description: str, is_warning: bool = False):
        """Scan Python files for specific patterns."""
        python_files = list(self.project_root.rglob('*.py'))
        
        # Exclude certain directories
        excluded_dirs = {'__pycache__', '.git', 'venv', 'env', 'node_modules', 'tests', 'testing', '.pytest_cache'}
        
        for file_path in python_files:
            # Skip files in excluded directories
            if any(excluded_dir in file_path.parts for excluded_dir in excluded_dirs):
                continue
            
            # Skip this validation script itself
            if file_path.name == 'validate_production.py':
                continue
            
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                    
                for pattern in patterns:
                    matches = re.finditer(pattern, content, re.IGNORECASE)
                    for match in matches:
                        line_num = content[:match.start()].count('\n') + 1
                        message = f"Found {description} in {file_path}:{line_num} - '{match.group()}'"
                        
                        if is_warning:
                            self.warnings.append(message)
                        else:
                            self.errors.append(message)
                            
            except Exception as e:
                self.warnings.append(f"Could not scan {file_path}: {str(e)}")


def main():
    """Main validation function."""
    validator = MiddlewareProductionValidator()
    results = validator.validate_all()
    
    print("\n" + "="*60)
    print("🏁 MIDDLEWARE PRODUCTION VALIDATION RESULTS")
    print("="*60)
    
    print(f"\n📊 Status: {results['status']}")
    print(f"🔍 Checks performed: {len(results['checks_performed'])}")
    
    if results['errors']:
        print(f"\n❌ Errors ({len(results['errors'])}):")
        for error in results['errors']:
            print(f"  • {error}")
    
    if results['warnings']:
        print(f"\n⚠️  Warnings ({len(results['warnings'])}):")
        for warning in results['warnings']:
            print(f"  • {warning}")
    
    if not results['errors'] and not results['warnings']:
        print("\n✅ All validation checks passed!")
    
    # Save results to file
    results_file = Path(__file__).parent / 'validation_results.json'
    with open(results_file, 'w') as f:
        json.dump(results, f, indent=2)
    
    print(f"\n📄 Detailed results saved to: {results_file}")
    
    # Exit with appropriate code
    sys.exit(0 if results['status'] == 'PASS' else 1)


if __name__ == '__main__':
    main()